### ISSUE TITLE
Validating dynamic map data

### ISSUE THREAD

**Issue by NathanBaulch** [go-ozzo/ozzo-validation#126]
I've been using ozzo-validation to validate typed API input for years but I recently found myself needing to validate dynamic `map[string]interface{}` data. I know, I *could* create typed models, marshal the data into them (from JSON or via mapstructure, etc) and validate them in the usual way but that's a lot of work/overhead in my particular situation.

Is there a recommended way to go about this?

In the meantime, I managed to get a reasonable solution working with the help of [go-cmp](https://github.com/google/go-cmp) which supports custom comparers. If the value being compared to implements `validation.Rule` then it uses the `Validate` method instead of comparing the values themselves.

```go
func main() {
	val := map[string]interface{}{
		"name":    "Nathan",
		"age":     33,
		"website": "https://nathan.dev",
	}
	exp := map[string]interface{}{
		"name":    validation.Required,
		"age":     validation.Min(18),
		"website": is.URL,
	}
	assertEqual(val, exp)
}
```

```go
func assertEqual(val, expect interface{}) {
	opt := cmp.FilterValues(func(x, y interface{}) bool {
		if _, ok := x.(validation.Rule); ok { return true }
		if _, ok := y.(validation.Rule); ok { return true }
		return false
	}, cmp.Comparer(func(x, y interface{}) bool {
		if r, ok := x.(validation.Rule); ok { return r.Validate(y) == nil }
		if r, ok := y.(validation.Rule); ok { return r.Validate(x) == nil }
		panic("unexpected")
	}))
	if diff := cmp.Diff(val, expect, opt, cmpopts.EquateErrors()); diff != "" {
		panic(diff)
	}
}
```

**Comment by qiangxue** [comment]
Your solution looks interesting! Validating a dynamic map could be a new feature to add. I'm thinking about the following usage syntax. What do you think?
```go
err := validation.ValidateMap(val,
    validation.Field("name", validation.Required, validation.Length(10, 20)),
    validation.Field("age", validation.Required, validation.Min(16)),
    validation.Field("website", is.URL),
)
```

**Comment by NathanBaulch** [comment]
Looks good. Would it be possible to have a strict option that disallows extra un-validated fields? My current `go-cmp` workaround does this by default since that package is designed to determine equality. This could also be an option for `ValidateStruct` if people want to be absolutely sure that they've considered all exported struct fields. Though I guess there would need to be a noop validation like `is.Anything` for fields that have no constraints.

Also, given that these maps are untyped, it would probably be useful to have `is.String`, `is.Integer`, etc to validate the type without any additional constraints. These could also have relaxed matching whereby "123" or float(123) are valid integers.

**Comment by qiangxue** [comment]
I like your suggestion of adding a strict option for both struct and map validation. Your suggestion of having type validation for untyped fields also makes sense.

**Comment by dlpetrie** [comment]
I'm in need of something like this as well. Would be very beneficial. The suggested syntax looks to be what I would expect.

**Comment by NathanBaulch** [comment]
I decided to give this feature a shot, example [here](https://github.com/NathanBaulch/ozzo-validation/blob/d29c7ac8cf0d06c87ddad0552b1cdf0897b7d2e7/example_test.go#L159-L189).

**Comment by dlpetrie** [comment]
That looks good to me!

**Comment by NathanBaulch** [comment]
Yep @dlpetrie, the latest proposal [here](https://github.com/NathanBaulch/ozzo-validation/blob/435b704280896849de1e3c7fc554a451a7d94bdb/example_test.go#L171-L187) supports this via the new `validation.Map` rule.

**Comment by kolkov** [comment]
This was resolved by @NathanBaulch's excellent PR #128 which added `validation.Map()` with `AllowExtraKeys()` support — exactly what was discussed here.

@dlpetrie — `Map()` works as a field-level rule within `ValidateStruct` too, so you can validate embedded map fields within structs.

Closing as implemented. Thanks @NathanBaulch for building this!
